### PR TITLE
[WIP] Add a higher level handling of flags

### PR DESCRIPTION
**Pull request by MatthieuDartiailh**
This is a work in progress on implementing #19. I am quite busy now so it may take a while to complete this I will ping when this is ready for review.

### CODE REVIEW

**Comment by MatthieuDartiailh** [comment]
I still need to add some docs but a first review would be welcome @haypo.

File: bytecode/concrete.py [hunk 1]
```diff
@@ -205,7 +204,7 @@ def from_code(code, *, extended_arg=False):
         bytecode = ConcreteBytecode()
         bytecode.name = code.co_name
         bytecode.filename = code.co_filename
-        bytecode.flags = code.co_flags
+        bytecode.flags = _bytecode.Flags(code.co_flags)
```
Comment (haypo) [inline bytecode/concrete.py]: Do you really need to explicitly convert to Flags? The setter does it for you, no? If not, please revert this change.

File: bytecode/concrete.py [hunk 2]
```diff
@@ -292,7 +291,7 @@ def to_code(self):
                               self.kwonlyargcount,
                               nlocals,
                               stacksize,
-                              self.flags,
+                              self.flags.to_int(self),
```
Comment (haypo) [inline bytecode/concrete.py]: Python has standard methods to cast to int: int(obj) calls obj.__int__() or obj.__index__().

File: bytecode/flags.py [hunk 3]
```diff
@@ -0,0 +1,297 @@
...
+class CoFlags(IntEnum):
+    """Possible values of the co_flags attribute of Code object.
+
+    Notes
+    -----
+    We do not rely on inspect values here as some of them are missing and
+    furthermore would be version dependent.
+
+    """
+    CO_OPTIMIZED             = 0x00001  # noqa
```
Comment (haypo) [inline bytecode/flags.py]: You might remove spaces rather than repeating #noqa, but it's up to you ;-)

File: bytecode/flags.py [hunk 4]
```diff
@@ -0,0 +1,297 @@
+# alias to keep the 'bytecode' variable free
+from enum import IntEnum
+from collections import defaultdict
+
+import bytecode as _bytecode
+
+
+class CoFlags(IntEnum):
```
Comment (haypo) [inline bytecode/flags.py]: Please rename to Flag.

File: bytecode/flags.py [hunk 5]
```diff
@@ -0,0 +1,297 @@
...
+    """Possible values of the co_flags attribute of Code object.
+
+    Notes
+    -----
+    We do not rely on inspect values here as some of them are missing and
+    furthermore would be version dependent.
```
Comment (haypo) [inline bytecode/flags.py]: I suggest to write: "Note: ...".

File: bytecode/flags.py [hunk 6]
```diff
@@ -0,0 +1,297 @@
...
+    object passed when converting to an integer. If no bytecode is provided the
+    default values are used.
+
+    The flags whose natural values can be inferred are:
+    - optimized
+    - generator
+    - nofree
+    - coroutine
+
+    """
```
Comment (haypo) [inline bytecode/flags.py]: I'm not a big fan of docstrings, it requires to duplicate the doc. The documentation in doc/ is more important than docstrings for me. You must document Flag and Flags in doc/. Then it's up to you to maintain docstrings as well, but in the project, I skipped docstrings to focus on good doc in doc/.

File: bytecode/flags.py [hunk 7]
```diff
@@ -0,0 +1,297 @@
...
+
+    def __init__(self, int_or_flags=0):
```
Comment (haypo) [inline bytecode/flags.py]: I suggest to rename the parameter to "value".

File: bytecode/flags.py [hunk 8]
```diff
@@ -0,0 +1,297 @@
...
+
+        self._forced = {}
+
+        if isinstance(int_or_flags, Flags):
+            self._defaults = int_or_flags._defaults.copy()
+            self._forced = int_or_flags._forced.copy()
+        elif isinstance(int_or_flags, int):
+            self._defaults = {name.split('_', 1)[1].lower():
+                              bool(int_or_flags & val)
+                              for name, val in CoFlags.__members__.items()}
```
Comment (haypo) [inline bytecode/flags.py]: I suggest to move "forced = {}" here.

File: bytecode/flags.py [hunk 9]
```diff
@@ -0,0 +1,297 @@
...
+        else:
+            msg = ("Flags object should be passed either None, an int  or a "
+                   "Flags instance at init, got {}.")
+            raise TypeError(msg.format(int_or_flags))
+
+    def __eq__(self, other):
+        if isinstance(other, int):
+            return self.to_int() == other
```
Comment (haypo) [inline bytecode/flags.py]: Hum, I would prefer to cast int to Flags, rather than the opposite. Would it work?

File: bytecode/flags.py [hunk 10]
```diff
@@ -0,0 +1,297 @@
...
+    - optimized
+    - generator
+    - nofree
+    - coroutine
+
+    """
+
+    def __init__(self, int_or_flags=0):
+
+        self._forced = {}
```
Comment (haypo) [inline bytecode/flags.py]: I don't understand the purpose of forced. Can you add a short comment here explaining its purpose?

File: bytecode/flags.py [hunk 11]
```diff
@@ -0,0 +1,297 @@
...
+    """Possible values of the co_flags attribute of Code object.
+
+    Notes
+    -----
+    We do not rely on inspect values here as some of them are missing and
+    furthermore would be version dependent.
+
+    """
+    CO_OPTIMIZED             = 0x00001  # noqa
+    CO_NEWLOCALS             = 0x00002  # noqa
```
Comment (haypo) [inline bytecode/flags.py]: I suggest to remove CO_ prefix.

File: bytecode/flags.py [hunk 12]
```diff
@@ -0,0 +1,297 @@
...
+    CO_VARARGS               = 0x00004  # noqa
```
Comment (haypo) [inline bytecode/flags.py]: I dislike hardcoded values. I would prefer to get them from a Python module, like dis.COMPILER_FLAG_NAMES, if possible.

File: bytecode/flags.py [hunk 13]
```diff
@@ -0,0 +1,297 @@
...
+    CO_VARKEYWORDS           = 0x00008  # noqa
+    CO_NESTED                = 0x00010  # noqa
+    CO_GENERATOR             = 0x00020  # noqa
+    CO_NOFREE                = 0x00040  # noqa
+    # New in Python 3.5
+    CO_COROUTINE             = 0x00080  # noqa
+    CO_ITERABLE_COROUTINE    = 0x00100  # noqa
```
Comment (haypo) [inline bytecode/flags.py]: Maybe only define them if the running Python version if >= 3.5? I don't know if bytecode has a policy to support a bytecode generated by a different version than the running Python.

File: bytecode/flags.py [hunk 14]
```diff
@@ -0,0 +1,297 @@
...
+        """The compiled code block has a varargs argument.
+
+        """
+        return self._forced.get('varargs', self._defaults['varargs'])
+
+    @varargs.setter
+    def varargs(self, value):
+        if not isinstance(value, bool):
+            raise ValueError('Flag value should be a boolean')
+        self._forced['varargs'] = value
```
Comment (haypo) [inline bytecode/flags.py]: Maybe you can factorize these methods with __getattr__() and __setattr__() methods, the code looks to be duplicated for each flag.

File: bytecode/tests/test_flags.py [hunk 15]
```diff
@@ -0,0 +1,103 @@
...
+
+class FlagsTests(unittest.TestCase):
+
+    def test_instantiation(self):
+
+        empty_flag = Flags()
+        self.assertFalse(any(empty_flag._defaults.values()))
+        self.assertFalse(empty_flag._forced)
+
+        flags = Flags(CoFlags.CO_OPTIMIZED + CoFlags.CO_NESTED)
```
Comment (haypo) [inline bytecode/tests/test_flags.py]: Hum, for flags, I prefer to write "a | b" than "a + b".

File: bytecode/tests/test_flags.py [hunk 16]
```diff
@@ -0,0 +1,103 @@
...
+        for m in CoFlags.__members__:
+            expected = (True
+                        if m in {'CO_OPTIMIZED', 'CO_NESTED'} else
+                        False)
+            self.assertEqual(flags.get_default(m.split('_', 1)[1].lower()),
```
Comment (haypo) [inline bytecode/tests/test_flags.py]: I would prefer a class method to get "all flags" than duplicating this split/lower code.

File: bytecode/tests/test_flags.py [hunk 17]
```diff
@@ -0,0 +1,103 @@
...
+        for val in CoFlags.__members__.values():
+            self.assertEqual(Flags(val).to_int(), val)
+
+        with self.assertRaises(ValueError):
+            Flags().to_int(Bytecode())
+
+        # Infer optimized and nofree
+        forced_flag = Flags()
+        forced_flag.optimized = False
+        forced_flag.nofree = False
```
Comment (haypo) [inline bytecode/tests/test_flags.py]: Please instanciate ConcreteBytecode here:

bytecode = ConcreteBytecode()

File: bytecode/flags.py [hunk 18]
```diff
@@ -0,0 +1,297 @@
...
+    CO_COROUTINE             = 0x00080  # noqa
+    CO_ITERABLE_COROUTINE    = 0x00100  # noqa
+    # New in Python 3.6
+    CO_ASYNC_GENERATOR       = 0x00200  # noqa
+
+    # __future__ flags
+    CO_FUTURE_GENERATOR_STOP = 0x80000  # noqa
+
+
+_CAN_DEDUCE_FROM_CODE = ('optimized', 'generator', 'nofree', 'coroutine')
```
Comment (haypo) [inline bytecode/flags.py]: "Can"? It seems like if you pass a concrete bytecode object, these flags are *always* set, no?

I'm asking if we cannot find a better name for the constant.

**Comment by haypo** [comment]
IMHO once we agree on the Flags design, maybe bytecode API would be complete enough for a version 1.0? What do you thing?

By the way, TODO.rst seems oudated: "ConcreteBytecode.to_code(): compute the stack level, see byteplay and Python/compile.c?" this item is now implemented, no?

Flags and stack depth were the two remaining blocker point to stabilize the API.

I'm not sure that bytecode has enough users to stabilize the API, but maybe it would have more users with a stable API? :-)

File: bytecode/concrete.py (hunk 1 above)
Comment (MatthieuDartiailh) [inline bytecode/concrete.py]: will do

File: bytecode/concrete.py (hunk 2 above)
Comment (MatthieuDartiailh) [inline bytecode/concrete.py]: The point here was to pass the bytecode to infer some flag values. If we want to rely on standard methods we need to set the bytecode as an attribute of the flag object.

File: bytecode/flags.py (hunk 4 above)
Comment (MatthieuDartiailh) [inline bytecode/flags.py]: will do

File: bytecode/flags.py (hunk 18 above)
Comment (MatthieuDartiailh) [inline bytecode/flags.py]: The point here is that the proper value of the flag can be deduced from the associated bytecode, so it is really a 'can'. 'Generator' and 'coroutine' are a bit special as they can be ambiguous (cf async_generator and iterable_coroutine)

File: bytecode/flags.py (hunk 6 above)
Comment (MatthieuDartiailh) [inline bytecode/flags.py]: I will add the docs in doc but I like docstrings so if it is fine with you I will leave them.

File: bytecode/flags.py (hunk 10 above)
Comment (MatthieuDartiailh) [inline bytecode/flags.py]: forced is used to keep track of flags whose values have been set by the user and whose value should hence not be inferred from the bytecode.

File: bytecode/flags.py (hunk 9 above)
Comment (MatthieuDartiailh) [inline bytecode/flags.py]: The issue is that if one value has been 'forcibly' set it will never match the int.

File: bytecode/flags.py (hunk 14 above)
Comment (MatthieuDartiailh) [inline bytecode/flags.py]: Will try to come up with something a bit less redundant yes.

File: bytecode/tests/test_flags.py (hunk 16 above)
Comment (MatthieuDartiailh) [inline bytecode/tests/test_flags.py]: make sense

File: bytecode/tests/test_flags.py (hunk 17 above)
Comment (MatthieuDartiailh) [inline bytecode/tests/test_flags.py]: Done

**Comment by MatthieuDartiailh** [comment]
This is superseded by #23, so closing
